pyvrp/cpp/DurationSegment.h

```diff
@@ -132,25 +137,48 @@ DurationSegment DurationSegment::merge(Matrix<Duration> const &durationMatrix,
     using Dur = pyvrp::Duration;
 
     // edgeDuration is the travel duration from our last to the other's first
-    // client, and atOther the time (after starting from our first client) at
-    // which we arrive there.
+    // client
     Dur const edgeDuration = durationMatrix(idxLast_, other.idxFirst_);
-    Dur const atOther = duration_ - timeWarp_ + edgeDuration;
 
-    // Time warp increases if we arrive at the other's first client after its
-    // time window closes, whereas wait duration increases if we arrive there
-    // before opening.
-    Dur const diffTw = std::max<Dur>(twEarly_ + atOther - other.twLate_, 0);
-    Dur const diffWait = other.twEarly_ - atOther > twLate_
-                             ? other.twEarly_ - atOther - twLate_
-                             : 0;  // ternary rather than max avoids underflow
+    // We must wait if we arrive before the earliest start time of the other
+    // segment. Note that the earliest start time of the other segment may not
+    // be the opening of the time window of the first client, but rather the
+    // time we should start to avoid unnecessary waiting time. We may still
+    // start earlier at that other segment, but this suggests that we should
+    // wait at another moment for the same total duration since we cannot
+    // finish the other route segment earlier.
+    Dur const waitDuration
+        = latestFinish_ < other.earliestStart_ - edgeDuration
+              ? other.earliestStart_ - latestFinish_ - edgeDuration
+              : 0;  // ternary rather than max avoids underflow
```

Comment:
For each case that has a ternary rather than a max: is this needed? I.e., does this cause over or underflow if we use max?

And for the cases that use max rather than a ternary: do those never overflow?

I wrote the cases out pretty carefully for the previous implementation, but I do not know yet what's needed for this proposal.

Reply:
Well it relies on the assumptions:
* That the edge duration is 'well behaved', i.e. never `max_value` but rather a large well behaved user value, which is checked and warned for by `Model.py`. If a user directly uses `ProblemData` this can give problems but I think the same is true for the current parameterization?
* That `earliestStart` is never infinite or a large value
* That `latestFinish`, replacing `twLate`, can be 'infinite'/`max_value`, which is why I use the ternary here to first check if that is the case.
* That `serviceDuration` for a client is a reasonable value and as such `duration` and `waitDuration` will all be reasonable values.

Summarizing, only `latestFinish` should be problematic as it can be 'infinite' which is handled explicitly, i.e. we either:
* subtract from `latestFinish` which should be ok
* we add to it if we have checked first the addition does not overflow
* we subtract it from something if we have checked first the result is positive

Regarding #590, I merged the test which needed to be slightly adapted as the input contained both timewarp and slack which is not feasible (I removed the slack by setting `tw_early = 18`). I validated that the modified test still reproduces the previous bug, but the proposed implementation passed the test without changes.